File: DataFeedViews/SlopeMultiplierDataFeedView.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports System.IO
Imports SyncroSim.Core
Imports SyncroSim.Core.Forms
Imports SyncroSim.Common.Forms
Imports System.Reflection

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class SlopeMultiplierDataFeedView

    Protected Overrides Sub InitializeView()

        MyBase.InitializeView()

        Dim v As DataFeedView =
            Me.Session.CreateMultiRowDataFeedView(Me.Scenario, Me.ControllingScenario)

        Me.PanelMultipliers.Controls.Add(v)

    End Sub

    Public Overrides Sub LoadDataFeed(dataFeed As Core.DataFeed)

        MyBase.LoadDataFeed(dataFeed)

        Dim v As DataFeedView = CType(Me.PanelMultipliers.Controls(0), DataFeedView)
        v.LoadDataFeed(dataFeed, DATASHEET_TRANSITION_SLOPE_MULTIPLIER_NAME)

    End Sub

    Public Overrides Sub EnableView(enable As Boolean)

        If (Me.PanelMultipliers.Controls.Count > 0) Then

            Dim v As DataFeedView = CType(Me.PanelMultipliers.Controls(0), DataFeedView)
            v.EnableView(enable)

        End If

        Me.TextBoxDEMFilename.Enabled = enable

        If (enable) Then
            Me.EnableButtons()
        Else
            Me.ButtonBrowse.Enabled = False
            Me.ButtonClear.Enabled = False
        End If

        Me.LabelDEM.Enabled = enable
        Me.LabelTMV.Enabled = enable

    End Sub

    Public Overrides Sub RefreshControls()

        MyBase.RefreshControls()

        Me.TextBoxDEMFilename.Text = Nothing

        Dim dr As DataRow = Me.GetDataRow()

        If (dr Is Nothing) Then
            Return
        End If

        Me.TextBoxDEMFilename.Text =
            CStr(dr(DATASHEET_DIGITAL_ELEVATION_MODEL_FILE_NAME_COLUMN_NAME))

    End Sub

    Private Sub EnableButtons()

        Me.ButtonBrowse.Enabled = True
        Me.ButtonClear.Enabled = False

        Dim dr As DataRow = Me.GetDataRow()

        If (dr Is Nothing) Then
            Return
        End If

        If (dr(DATASHEET_DIGITAL_ELEVATION_MODEL_FILE_NAME_COLUMN_NAME) IsNot DBNull.Value) Then

            Me.ButtonBrowse.Enabled = False
            Me.ButtonClear.Enabled = True

        End If

    End Sub

    Private Function GetDataSheet() As DataSheet

        If (Me.DataFeed IsNot Nothing) Then
            Return Me.DataFeed.GetDataSheet(DATASHEET_DIGITAL_ELEVATION_MODEL_NAME)
        Else
            Return Nothing
        End If

    End Function

    Private Function GetDataRow() As DataRow

        Dim ds As DataSheet = Me.GetDataSheet()

        If (ds IsNot Nothing) Then
            Return ds.GetDataRow()
        Else
            Return Nothing
        End If

    End Function

    Private Sub ButtonBrowse_Click(sender As System.Object, e As System.EventArgs) Handles ButtonBrowse.Click

        Dim RasterFile As String = ChooseRasterFileName("Digital Elevation Model File")

        If (RasterFile Is Nothing) Then
            Return
        End If

        Using h As New HourGlass

            Dim ds As DataSheet = Me.GetDataSheet()
            Dim dr As DataRow = ds.GetDataRow()
            Dim RasterFileName As String = Path.GetFileName(RasterFile)

            If (dr Is Nothing) Then

                dr = ds.GetData.NewRow()

                ds.BeginAddRows()
                dr(DATASHEET_DIGITAL_ELEVATION_MODEL_FILE_NAME_COLUMN_NAME) = RasterFileName
                ds.GetData.Rows.Add(dr)
                ds.EndAddRows()

            Else

                ds.BeginModifyRows()
                dr(DATASHEET_DIGITAL_ELEVATION_MODEL_FILE_NAME_COLUMN_NAME) = RasterFileName
                ds.EndModifyRows()

            End If

            ds.AddExternalInputFile(RasterFile)

            Me.RefreshControls()
            Me.EnableButtons()

        End Using

    End Sub

    Private Sub ButtonClear_Click(sender As System.Object, e As System.EventArgs) Handles ButtonClear.Click

        Using h As New HourGlass

            Dim ds As DataSheet = GetDataSheet()

            ds.ClearData()
            Me.RefreshControls()
            Me.EnableButtons()

        End Using

    End Sub

End Class

File: src/DataFeedViews/SlopeMultiplierDataFeedView.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2024 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using System.IO;
using System.Data;
using System.Globalization;
using SyncroSim.Core;
using SyncroSim.Core.Forms;
using SyncroSim.Apex.Forms;

namespace SyncroSim.STSim
{
    internal partial class SlopeMultiplierDataFeedView
    {
        public SlopeMultiplierDataFeedView()
        {
            InitializeComponent();
        }

        protected override void InitializeView()
        {
            base.InitializeView();

            DataFeedView v = this.Session.CreateMultiRowDataFeedView(this.Scenario, this.ControllingScenario);

            this.PanelMultipliers.Controls.Add(v);
        }

        public override void LoadDataFeed(Core.DataFeed dataFeed)
        {
            base.LoadDataFeed(dataFeed);

            DataFeedView v = (DataFeedView)this.PanelMultipliers.Controls[0];
            v.LoadDataFeed(dataFeed, Strings.DATASHEET_TRANSITION_SLOPE_MULTIPLIER_NAME);
        }

        public override void EnableView(bool enable)
        {
            if (this.PanelMultipliers.Controls.Count > 0)
            {
                DataFeedView v = (DataFeedView)this.PanelMultipliers.Controls[0];
                v.EnableView(enable);
            }

            this.TextBoxDEMFilename.Enabled = enable;

            if (enable)
            {
                this.EnableButtons();
            }
            else
            {
                this.ButtonBrowse.Enabled = false;
                this.ButtonClear.Enabled = false;
            }

            this.LabelDEM.Enabled = enable;
            this.LabelTMV.Enabled = enable;
        }

        public override void RefreshControls()
        {
            base.RefreshControls();

            this.TextBoxDEMFilename.Text = null;

            DataRow dr = this.GetDataRow();

            if (dr == null)
            {
                return;
            }

            this.TextBoxDEMFilename.Text = Convert.ToString(
                dr[Strings.DATASHEET_DIGITAL_ELEVATION_MODEL_FILE_NAME_COLUMN_NAME], 
                CultureInfo.InvariantCulture);
        }

        private void EnableButtons()
        {
            this.ButtonBrowse.Enabled = true;
            this.ButtonClear.Enabled = false;

            DataRow dr = this.GetDataRow();

            if (dr == null)
            {
                return;
            }

            if (dr[Strings.DATASHEET_DIGITAL_ELEVATION_MODEL_FILE_NAME_COLUMN_NAME] != DBNull.Value)
            {
                this.ButtonBrowse.Enabled = false;
                this.ButtonClear.Enabled = true;
            }
        }

        private DataSheet GetDataSheet()
        {
            if (this.DataFeed != null)
            {
                return this.DataFeed.GetDataSheet(Strings.DATASHEET_DIGITAL_ELEVATION_MODEL_NAME);
            }
            else
            {
                return null;
            }
        }

        private DataRow GetDataRow()
        {
            DataSheet ds = this.GetDataSheet();

            if (ds != null)
            {
                return ds.GetDataRow();
            }
            else
            {
                return null;
            }
        }

        private void ButtonBrowse_Click(object sender, System.EventArgs e)
        {
            string RasterFile = RasterUtilities.ChooseRasterFileName("Digital Elevation Model File", this);

            if (RasterFile == null)
            {
                return;
            }

            using (HourGlass h = new HourGlass())
            {
                DataSheet ds = this.GetDataSheet();
                DataRow dr = ds.GetDataRow();
                string RasterFileName = Path.GetFileName(RasterFile);

                if (dr == null)
                {
                    dr = ds.GetData().NewRow();

                    ds.BeginAddRows();
                    dr[Strings.DATASHEET_DIGITAL_ELEVATION_MODEL_FILE_NAME_COLUMN_NAME] = RasterFileName;
                    ds.GetData().Rows.Add(dr);
                    ds.EndAddRows();
                }
                else
                {
                    ds.BeginModifyRows();
                    dr[Strings.DATASHEET_DIGITAL_ELEVATION_MODEL_FILE_NAME_COLUMN_NAME] = RasterFileName;
                    ds.EndModifyRows();
                }

                ds.AddExternalInputFile(RasterFile);

                this.RefreshControls();
                this.EnableButtons();
            }
        }

        private void ButtonClear_Click(object sender, System.EventArgs e)
        {
            using (HourGlass h = new HourGlass())
            {
                DataSheet ds = GetDataSheet();

                ds.ClearData();
                this.RefreshControls();
                this.EnableButtons();
            }
        }
    }
}
